README
$Id: README,v 1.4 2006/11/01 22:43:33 mike Exp $

Games::ScottAdams - utilities for manipulating Scott Adams adventures


INTRODUCTION
------------

Games::ScottAdams is a Perl module for manipulating adventure games in
the very obsolete but strangely compelling Scott Adams format.  In
particular, it includes "sac", the Scott Adams Compiler, which
compiles a human-readable description of an adventure into a format
that can be interpreted by common run-time systems such as scottfree.
This provides a nice, low-overhead way in for people who want to start
building their own adventures.

Aside from the obvious Perl-module stuff (MANIFEST, Makefile.PL,
ScottAdams.pm and the ScottAdams subdirectory for sub-modules and
test.pl), what's what in this distribution?

README		This file.
GPL-2		The licence under which this software is distributed.
Changes		History of the distribution, and things still to do.
sac		The "wrapper" program that invokes the module.
dubbin/		Uninteresting files used in regression-testing.
crystal/	_Crystal of Chaos_, a sample adventure built with sac.
tutorial/	The bits and pieces from which Tutorial.pm is built.
		Change into this directory, run "make", and view the
		generated HTML file "Tutorial.html"

You may wish to avoid reading the _Crystal of Chaos_ source code
(crystal/crystal.sac) until you've solved the game.


SYNOPSIS
--------

	$ perl Makefile.PL
	$ make
	$ make test
	$ sudo make install
	$ cd crystal
	$ sac crystal.sac > crystal.sao
	$ scottfree crystal.sao


AUTHOR
------

All code and documentation by Mike Taylor <mike@miketaylor.org.uk>
	http://www.miketaylor.org.uk/
Please email me with bug-reports, wishlist items, patches, deployment
stories and, of course, large cash donations.


LICENCE
-------

The Games::ScottAdams module is Free Software, which is pretty much
legally equivalent -- though not morally equivalent -- to Open Source.
See
	http://www.gnu.org/philosophy/free-software-for-freedom.html
for a detailed if somewhat one-sided discussion of the differences,
and particularly of why Free Software is an important idea.

Games::ScottAdams is distributed under version 2 of the GPL (GNU
GENERAL PUBLIC LICENSE) of June 1991.  A copy of the licence is
included in this distribution, as the file GPL-2.  This licence does
not allow you to restrict the freedom of others to use derived
versions of Games::ScottAdams (i.e. you must share your enhancements),
nor to deploy Games::ScottAdams as a part of a non-free larger work.


SEE ALSO
--------

The Games::ScottAdams tutorial:
	lib/Games/ScottAdams/Tuturial.pod

The Games::ScottAdams reference manual:
	lib/Games/ScottAdams/Manual.pod

The "Changes" file, including the "Still to do" section.

ScottCom and other tools that act as compilers and decompilers for
various Scott-Adams adventure languages:
	http://ifarchive.flavorplex.com/if-archive/scott-adams/tools/?S=A

ScottFree, a nice interpreter that uses sac's compiled files.  This is
included as a standard package in many Linux distributions, including
Debian and Ubuntu.


Changes
Revision history for Perl extension Games::ScottAdams.

0.03  Thu Nov  2 00:30:45 GMT 2006
	- Revised directory layout to match that produced by h2xs
	  version 1.23 with flags:
		-X --name=Games::ScottAdams --compat-version=5.8.0 --omit-constant --skip-exporter --skip-ppport
	- Fix occasional bug in %nalias: the first of the equivalent
	  nouns to appear in the vocabulary list is now always the one
	  that is used as the %getdrop, so that it is recognised by
	  scottfree.
	- Finish Games::ScottAdams::Manual
	- Remove extraneous warning emitted when %AT places an object
	  that already had a location.
	- The compiler is not now so fussy about leading whitespace.
	  In general, all indentation is now ignored except when it's
	  part of a location description.
	- Move the sac executable down into new bin directory.
	- Use ".sao" (Scott Adams Object) instead of the uninformative
	  ".dat" throughout as the suffix for compiled files.
	- Tweaks in crystal.sac
	- Fix crystal/Makefile to use installed components.

0.02  Mon Jun  7 12:49:50 2004
	- First release.

0.01  Tue Apr 17 13:42:46 2001
	- original version; created by h2xs 1.19

--

### Still to do
	- Allow room/item descriptions on %whatever line.
	- Naming and reusing of messages.
	- Handle multi-line messages.
	- Add meta-conditions such as "dark" that are implemented in
	  terms of primitive flags.
	- Mechanism for setting and testing flags by name.
	  (flag 15 always indicates if light has run out.)
	- Register current line-numbers with rooms, actions, etc. so
	  that compile-phase errors can be reported better.
	- Add discussion of %ident, %version, %wordlen, %maxload,
	  %lighttime and %include, and flags, counters and location
	  stores, to the tutorial.
	- Test all the action conditions and results.
	- Write decompiler; regression-test against compiler.
	- Figure out the MakeMaker magic to have
	  lib/Games/ScottAdams/Tuturial.pod automatically built.

It's possible to write perfectly good games (e.g. to clone Scott
Adams's own _Adventureland_) without any of these refinements.


lib/Games/ScottAdams.pm
# $Id: ScottAdams.pm,v 1.3 2006/11/01 22:47:57 mike Exp $

# ScottAdams.pm - a compiler for Scott Adams adventures

package Games::ScottAdams;
use strict;

our $VERSION = '0.03';

=head1 NAME

Games::ScottAdams - Perl extension for representing Scott Adams games

=head1 SYNOPSIS

	use Games::ScottAdams;
	$game = new Games::ScottAdams::Game();
	$game->parse('/usr/local/lib/sac/foo.sac');
	$game->compile();

=head1 DESCRIPTION

This module allows adventure games in the textual SAC format to be
compiled into a form that can be understood by Scott Adams adventure
interpreters such as ScottFree and GnomeScott.

I don't propose to discuss the API because frankly, no-one will ever
call it.  They'll just use the trivial front-end program C<sac> which
is essentially identical to the code in the SYNOPSIS, but with a few
more C<use strict>s and suchlike.

You'd do much better to read the Tutorial and Reference Manual.


=head1 SEE ALSO

I<The Scott Adams Adventure Compiler Tutorial>
(B<Games::ScottAdams::Tutorial>)

I<The Scott Adams Adventure Compiler Reference Manual>
(B<Games::ScottAdams::Manual>)

C<sac>, the Scott Adams Compiler that uses this module.

=head1 AUTHOR

Mike Taylor E<lt>mike@miketaylor.org.ukE<gt>

First version Tuesday 17th April 2001.

=cut


use Games::ScottAdams::Game;

1;
__END__


lib/Games/ScottAdams/File.pm
# $Id: File.pm,v 1.1 2006/10/31 20:31:21 mike Exp $

# File.pm - a cleverer IO::File-alike that does pushback

package Games::ScottAdams::File;
use strict;

# This module simply implements a slightly cleverer IO::File-alike
# that remembers the filename, maintains a notion of the current line
# number (useful for diagnostics) and can maintain an arbitrary number
# of pushback lines.  It clearly has wider applicability outside of
# the Scott Adams module and should probably not be a
# Games::ScottAdams class.

use IO::File;


sub new {
    my $class = shift();
    my($filename) = @_;

    my $f = new IO::File("<$filename")
	or return undef;

    return bless {
	f => $f,
	filename => $filename,
	linenumber => 0,
	pushback => [],
    }, $class;
}


sub getline {
    my $this = shift();
    my($trim) = @_;

    my $line = pop @{ $this->{pushback} };
    if (!defined $line) {
      AGAIN:
	$this->{linenumber}++;
	$line = $this->{f}->getline();
	return undef if !defined $line;
    }

    if ($trim) {
	$line =~ s/#.*//;
	$line =~ s/\s+$//;
	goto AGAIN if $line =~ /^$/;
    }

    return $line;
}


sub ungetline {
    my $this = shift();
    my($line) = @_;

    push @{ $this->{pushback} }, $line;
}


sub warn {
    my $this = shift();

    print STDERR $this->{filename}, ':', $this->{linenumber}, ': ',
	'WARNING: ', @_, "\n";
}


sub fatal {
    my $this = shift();

    my $filename = $this->{filename} || '[unknown]';
    my $linenumber = $this->{linenumber} || '[unknown]';
    print STDERR $filename, ':', $linenumber, ': ERROR: ', @_, "\n";
    exit 1;
}


sub close {
    my $this = shift();

    $this->{f}->close();
}


1;


lib/Games/ScottAdams/Action.pm
# $Id: Action.pm,v 1.3 2006/11/01 22:59:52 mike Exp $

# Action.pm - an action in a Scott Adams game.

package Games::ScottAdams::Action;
use strict;


sub new {
    my $class = shift();
    my($verb, $noun, $num) = @_;

    return bless {
	verb => $verb,
	noun => $noun,
	num => $num,		# 0-based index into Game's list of actions
				### I don't think we actually use this
	comment => undef,	# optional comment to be written through
	cond => [],		# array of conditions to be satisfied
	res => [],		# array of results to be executed
    }, $class;
}


sub verb {
    my $this = shift();
    return $this->{verb};
}

sub noun {
    my $this = shift();
    return $this->{noun};
}

sub comment {
    my $this = shift();
    my($name) = @_;

    my $old = $this->{comment};
    if (defined $name) {
	$this->{comment} = $name;
    }
    return $old;
}


# We'd like to compile these up front so we can complain about
# unrecognised condition and actions while we still know where we are
# in the source file.  Unfortunately, we can't do it in general as the
# action may refer to the names of rooms or items that have not yet
# been defined.  So all we can do at this stage is remember them for
# later.
#
sub add_cond {
    my $this = shift();
    my($text) = @_;

    push @{ $this->{cond} }, $text;
}


sub add_result {
    my $this = shift();
    my($text) = @_;

    push @{ $this->{res} }, $text;
}


# PRIVATE to the compile() method.
sub ARG_NONE { 0 }		# no argument
sub ARG_NUM { 1 }		# argument specifies a flag
sub ARG_ROOM { 2 }		# argument identifies a room
sub ARG_ITEM { 3 }		# argument identifies an item
sub ARG_ITEMROOM { 4 }		# arguments identify an item and a room
sub ARG_ITEMITEM { 5 }		# arguments identify two items

my %_cond = (
	     carried =>		[ 1, ARG_ITEM ],
	     here =>		[ 2, ARG_ITEM ],
	     accessible =>	[ 3, ARG_ITEM ],
	     at =>		[ 4, ARG_ROOM ],
	     '!here' =>		[ 5, ARG_ITEM ],
	     '!carried' =>	[ 6, ARG_ITEM ],
	     '!at' =>		[ 7, ARG_ROOM ],
	     flag =>		[ 8, ARG_NUM ],
	     '!flag' =>		[ 9, ARG_NUM ],
	     loaded =>		[ 10, ARG_NONE ],
	     '!loaded' =>	[ 11, ARG_NONE ],
	     '!accessible' =>	[ 12, ARG_ITEM ],
	     exists =>		[ 13, ARG_ITEM ],
	     '!exists' =>	[ 14, ARG_ITEM ],
	     counter_le =>	[ 15, ARG_NUM ],
#	     counter_ge =>	[ 16, ARG_NUM ],
	     counter_gt =>	[ 16, ARG_NUM ],
		#   ###	The documentation accompanying the scottfree
		#	interpreter says that condition 16 tests for
		#	current counter's value greater than or equal
		#	to the argument, but inspection of the source
		#	shows that it actually tests for strictly
		#	greater-than.
	     '!moved' =>	[ 17, ARG_ITEM ],
	     moved =>		[ 18, ARG_ITEM ],
	     counter_eq =>	[ 19, ARG_NUM ],
	     );

my %_res = (
	    get =>		[ 52, ARG_ITEM ],
	    drop =>		[ 53, ARG_ITEM ],
	    moveto =>		[ 54, ARG_ROOM ],
	    destroy =>		[ 55, ARG_ITEM ],
	    set_dark =>		[ 56, ARG_NONE ],
	    clear_dark =>	[ 57, ARG_NONE ],
	    set_flag =>		[ 58, ARG_NUM ],
	    destroy2 =>		[ 59, ARG_ITEM ],
		# Same as 55 in ScottFree
	    clear_flag =>	[ 60, ARG_NUM ],
	    die =>		[ 61, ARG_NONE ],
	    put =>		[ 62, ARG_ITEMROOM ],
	    game_over =>	[ 63, ARG_NONE ],
	    look =>		[ 64, ARG_NONE ],
	    score =>		[ 65, ARG_NONE ],
	    inventory =>	[ 66, ARG_NONE ],
	    set_0 =>		[ 67, ARG_NONE ],
	    clear_0 =>		[ 68, ARG_NONE ],
	    refill_lamp =>	[ 69, ARG_NONE ],		### UNTESTED
	    clear_screen =>	[ 70, ARG_NONE ],		### UNTESTED
	    save_game =>	[ 71, ARG_NONE ],
	    swap =>		[ 72, ARG_ITEMITEM ],
	    continue =>		[ 73, ARG_NONE ],		### UNTESTED
		# Automatic -- is there ever any need to use it explicitly?
	    superget =>		[ 74, ARG_ITEM ],		### UNTESTED
	    put_with =>		[ 75, ARG_ITEMITEM ],
	    look2 =>		[ 76, ARG_NONE ],		### UNTESTED
		# Same as 64 in ScottFree
	    decrease_counter =>	[ 77, ARG_NONE ],
	    print_counter =>	[ 78, ARG_NONE ],
	    set_counter =>	[ 79, ARG_NUM ],
	    swap_loc_default =>	[ 80, ARG_NONE ],
	    select_counter =>	[ 81, ARG_NUM ],		### UNTESTED
		# Current counter is swapped with specified backup counter
	    add_counter =>	[ 82, ARG_NUM ],		### UNTESTED
	    subtract_counter =>	[ 83, ARG_NUM ],		### UNTESTED
	    print_noun =>	[ 84, ARG_NONE ],
	    print_noun_nl =>	[ 85, ARG_NONE ],
	    nl =>		[ 86, ARG_NONE ],
	    swap_loc =>		[ 87, ARG_NUM ],
	    pause =>		[ 88, ARG_NONE ],
	    special =>		[ 89, ARG_NUM ],
		# This is special -- see ../../../../scottfree/Definition
	    );


sub compile {
    my $this = shift();
    my($game) = @_;

    my $verb = $game->resolve_verb($this->verb());
    my $noun = $this->noun();
    if ($verb == 0) {
	# This is a %occur, so the noun is a percentage probability
	$noun = 100 if !$noun;
    } else {
	$noun = $game->resolve_noun($noun);
    }

    my @condval = ( 150*$verb + $noun );
    foreach my $cond (@{ $this->{cond} }) {
	my($opcode, $arg) = _lookup($game, $cond, 'condition', \%_cond);
	$arg = 0 if !defined $arg;
	push @condval, $opcode + 20*$arg;
    }

    die "Oops!  SA format doesn't support >5 conditions in an action"
	if @condval > 6;

    # Now gather results, with parameters going on the end of @condval
    #warn "handling results:\n" . join ('', map {"\t$_\n"}
    #				       @{ $this->{res} });
    my @resval;
    foreach my $res (@{ $this->{res} }) {
	my($opcode, @arg) = _lookup($game, $res, 'result', \%_res);
	push @resval, [ $opcode, @arg ];
    }

    # Right.  This is slightly tricky.  We now want to pack all the
    # results, together with their parameters, into as few action
    # octuplets as possible.  We have four result slots available in
    # the first one, together with zero or more parameter slots
    # remaining in the condition area; thereafter, each action
    # octuplet offers four more result slots together with five
    # parameter slots in the condition area (which of course is one
    # more than we'll ever need.)
    my @conds;			# list of completed octuplets
    my $argslot = @condval;	# 0-based index within current octuplet
    my $resslot = 0;		# 0-based index into "virtual array"
    push @condval, map { 0 } 1..(8-@condval);

    for (my $i = 0; $i < @resval; $i++) {
	my $res = $resval[$i];
	my($opcode, @arg) = @$res;
	@arg = grep { defined } @arg;

	### Seems like 6 in next line should be 5.  Think harder.
	if ($argslot + @arg > 6 || $resslot == 4 ||
	    ($resslot == 3 && $i < @resval-1)) {
	    # Current octuplet is full: skip to next
	    my $cindex = 6 + int($resslot/2);
	    $condval[$cindex] +=
		($resslot % 2 == 0 ? 150 : 1) * 73;
	    push @conds, join(' ', @condval);
	    @condval = map { 0 } 1..8;
	    $argslot = 1;	# because slot 0 holds verb & noun
	    $resslot = 0;
	}

	my $cindex = 6 + int($resslot/2);
	$condval[$cindex] +=
	    ($resslot % 2 == 0 ? 150 : 1) * $opcode;
	$resslot++;
	foreach my $arg (@arg) {
	    if (!defined $arg) {
		print STDERR "", "arg in '@arg' (", scalar(@arg), ") undef\n";
	    }
	    $condval[$argslot] = 20*$arg;
	    $argslot++;
	}
    }

    push @conds, join(' ', @condval);
#print STDERR "", "returning conds: ", join(' -- ', @conds), "\n";
    return @conds;
}


sub _lookup {
    my($game, $text, $caption, $href) = @_;

    $text =~ s/^\s+//;
    my($op, $arg) = split /\s+/, $text, 2;
    if ($op eq 'msg') {
	# This check is a hack, but does spot an otherwise subtle bug
	die "Oops!  `msg' used as a condition (missing %result line?)"
	    if $caption eq 'condition';

	my $mnum = $game->resolve_message($arg);
	return ($mnum <= 51 ? $mnum : $mnum+51);
    }

    my $ref = $href->{$op};
    die "unrecognised $caption op '$op'"
	if !defined $ref;

    my($opcode, $argtype) = @$ref;
    if ($argtype == ARG_NONE) {
	return ($opcode);
    } elsif ($argtype == ARG_NUM) {
	# Numeric argument already has the right numeric value.
    } elsif ($argtype == ARG_ROOM) {
	$arg = $game->resolve_room($arg, 'action');
    } elsif ($argtype == ARG_ITEM) {
	$arg = $game->resolve_item($arg, 'action');
    } elsif ($argtype == ARG_ITEMROOM) {
	my($arg1, $arg2) = split /\s+/, $arg, 2;
	$arg1 = $game->resolve_item($arg1, 'action');
	$arg2 = $game->resolve_room($arg2, 'action');
	return ($opcode, $arg1, $arg2);
    } elsif ($argtype == ARG_ITEMITEM) {
	my($arg1, $arg2) = split /\s+/, $arg, 2;
	$arg1 = $game->resolve_item($arg1, 'action');
	$arg2 = $game->resolve_item($arg2, 'action');
	return ($opcode, $arg1, $arg2);
    } else {
	die "unsupported argument type $argtype for op '$op'";
    }

    return ($opcode, $arg);
}


1;


lib/Games/ScottAdams/Parse.pm
# $Id: Parse.pm,v 1.2 2006/10/31 21:51:43 mike Exp $

# Parse.pm - parsing functions for Scott Adams game files.

package Games::ScottAdams::Game;
use strict;

use Games::ScottAdams::File;


sub parse {
    my $this = shift();
    my($filename) = @_;

    #warn "parsing '$filename'";
    my $fh = new Games::ScottAdams::File($filename)
	or die "can't open '$filename': $!";

    while (defined ($_ = $fh->getline(1))) {
	s/^\s+%/%/;		# Skip any whitespace before leading %
	if (/^%room (.*)/i) {
	    $this->_parse_room($fh, $1);
	} elsif (/^%exit\s+(.*)\s+(.*)/i) {
	    $this->_parse_exit($fh, $1, $2);
	} elsif (/^%item\s+(.*)/i) {
	    $this->_parse_item($fh, $1);
	} elsif (/^%getdrop\s+(.*)/i) {
	    $this->_parse_getdrop($fh, $1);
	} elsif (/^%(at|nowhere)\s*(.*)/i) {
	    $this->_parse_at($fh, $2);
	} elsif (/^%start\s+(.*)/i) {
	    $this->_parse_start($fh, $1);
	} elsif (/^%treasury\s+(.*)/i) {
	    $this->_parse_treasury($fh, $1);
	} elsif (/^%maxload\s+(.*)/i) {
	    $this->_parse_maxload($fh, $1);
	} elsif (/^%lighttime\s+(.*)/i) {
	    $this->_parse_lighttime($fh, $1);
	} elsif (/^%ident\s+(.*)/i) {
	    $this->_parse_ident($fh, $1);
	} elsif (/^%version\s+(.*)/i) {
	    $this->_parse_version($fh, $1);
	} elsif (/^%wordlen\s+(.*)/i) {
	    $this->_parse_wordlen($fh, $1);
	} elsif (/^%lightsource\s+(.*)/i) {
	    $this->_parse_lightsource($fh, $1);
	} elsif (/^%action\s+(.*)\s+(.*)/i) {
	    $this->_parse_action($fh, $1, $2);
	} elsif (/^%action\s+(.*)/i) {
	    $this->_parse_action($fh, $1, undef);
	} elsif (/^%occur\s*(.*)/i) {
	    $this->_parse_action($fh, undef, $1);
	} elsif (/^%result/i) {
	    $this->_parse_result($fh);
	} elsif (/^%comment\s+(.*)/i) {
	    $this->_parse_comment($fh, $1);
	} elsif (/^%([nv])alias\s+(.*)\s+(.*)/i) {
	    $this->_parse_alias($fh, lc($1), $2, $3);
	} elsif (/^%include\s+(.*)/i) {
	    my $newfile = $1;
	    #warn "%include '$newfile'";
	    # Interpret filenames relative to directory of current file
	    if ($filename =~ m@/@) {
		my $prefix = $filename;
		$prefix =~ s@(.*)/.*@$1@;
		$newfile = "$prefix/$newfile";
	    }
	    $this->parse($newfile);
	} elsif (!/^%/) {
	    $fh->warn("expected directive, got '$_' (ignored)");
	} else {
	    $fh->warn("unrecognised directive (ignored): '$_'");
	}
    }

    $this->_coalesce_aliases();
    return 1;
}


# PRIVATE to the parse() method
sub _parse_room {
    my $this = shift();
    my($fh, $name) = @_;

    my $desc = '';
    while (defined (my $line = $fh->getline(1))) {
	if ($line =~ /^\s*%/) {
	    $fh->ungetline($line);
	    last;
	}

	$desc .= "$line\n";
    }

    my $num = @{ $this->{rooms} }; # 0-based index of room to be added
    my $room = new Games::ScottAdams::Room($name, $desc, $num);
    push @{ $this->{rooms} }, $room;
    if (defined $this->{roomname}->{$name}) {
	$fh->warn("discarding old room '$name'");
    }

    $this->{roomname}->{$name} = $room;
    $this->{_room} = $room;
    $this->{_roomname1} = $name
	if !defined $this->{_roomname1};
}


# PRIVATE to the parse() method
sub _parse_exit {
    my $this = shift();
    my($fh, $dir, $dest) = @_;

    my $room = $this->{_room};
    if (!defined $room) {
	$fh->warn("ignoring %exit '$dir'->'$dest' before first room");
	return;
    }

    my $roomname = $room->name();
    $dir = lc($dir);
    if ($dir !~ /^[nsewud]$/) {
	$fh->warn("ignoring %exit '$dir'->'$dest' at '$roomname'");
	return;
    }

    my $old = $room->exit($dir);
    if (defined $old) {
	$fh->warn("discarding old exit '$dir'->'$old' at '$roomname'");
    }

    $room->exit($dir, $dest);
}


# PRIVATE to the parse() method
sub _parse_item {
    my $this = shift();
    my($fh, $name) = @_;

    my $where = undef;		# item is initially nowhere
    my $room = $this->{_room};
    if (defined $room) {
	$where = $room->name();
    }

    my $desc = $fh->getline(1);
    my $num = @{ $this->{items} }; # 0-based index of item to be added
    if (0) {
	### No need to do this, is there?
	if ($num == 9) {
	    # Leave slot 9 free for the light-source
	    my $nothing = new Games::ScottAdams::Item('', '', $num++);
	    push @{ $this->{items} }, $nothing;
	}
    }

    my $item = new Games::ScottAdams::Item($name, $desc, $num, $where);
    push @{ $this->{items} }, $item;
    if (defined $this->{itemname}->{$name}) {
	$fh->warn("discarding old item '$name'");
    }

    $this->{itemname}->{$name} = $item;
    $this->{_item} = $item;
    if ($desc =~ /^\*/) {
	$this->{ntreasures}++;
    }
}


# PRIVATE to the parse() method
sub _parse_getdrop {
    my $this = shift();
    my($fh, $name) = @_;

    my $item = $this->{_item};
    if (!defined $item) {
	$fh->warn("ignoring %getdrop '$name' before first item");
	return;
    }

    my $itemname = $item->name();
    my $old = $item->getdrop();
    if (defined $old) {
	$fh->warn("discarding old getdrop '$old' for '$itemname'");
    }

    $item->getdrop($name);
    $this->_parse_alias($fh, 'n', $name);
}


# PRIVATE to the parse() method
sub _parse_at {
    my $this = shift();
    my($fh, $where) = @_;

    my $item = $this->{_item};
    if (!defined $item) {
	$fh->warn("ignoring %at '$where' before first item");
	return;
    }

    my $itemname = $item->name();
    my $old = $item->where();
    if (defined $old && $where ne '') {
	#$fh->warn("replacing location '$old' with '$where' for '$itemname'");
    }

    $item->where($where);
}


# All the following wrappers are PRIVATE to the parse() method
sub _parse_start {
    return _parse_param(@_, 'start', 'start room'); }
sub _parse_treasury {
    return _parse_param(@_, 'treasury', 'treasury room'); }
sub _parse_maxload {
    return _parse_param(@_, 'maxload', 'maximum load'); }
sub _parse_lighttime {
    return _parse_param(@_, 'lighttime', 'light duration'); }
sub _parse_ident {
    return _parse_param(@_, 'ident', 'adventure identifier'); }
sub _parse_version {
    return _parse_param(@_, 'version', 'version number'); }
sub _parse_wordlen {
    return _parse_param(@_, 'wordlen', 'word length'); }
sub _parse_lightsource {
    return _parse_param(@_, 'lightsource', 'light source'); }


# PRIVATE to the _parse_{start...wordlen}() methods
sub _parse_param {
    my $this = shift();
    my($fh, $value, $param, $caption) = @_;

    if (defined $this->{$param}) {
	$fh->warn("discarding old $caption '", $this->{$param}, "'");
    }

    $this->{$param} = $value;
}


# PRIVATE to the parse() method
sub _parse_action {
    my $this = shift();
    my($fh, $verb, $noun) = @_;

    my $num = @{ $this->{actions} }; # 0-based index of action to be added
    my $action = new Games::ScottAdams::Action($verb, $noun, $num);
    push @{ $this->{actions} }, $action;
    $this->{_action} = $action;

    # Register noun and verb
    $this->_parse_alias($fh, 'v', $verb) if $verb;
    $this->_parse_alias($fh, 'n', $noun) if $noun;

    while (defined (my $line = $fh->getline(1))) {
	if ($line =~ /^\s*%/) {
	    $fh->ungetline($line);
	    last;
	}
	$action->add_cond($line);
    }

    return;
}


# PRIVATE to the parse() method
sub _parse_result {
    my $this = shift();
    my($fh) = @_;

    my $action = $this->{_action};
    if (!defined $action) {
	$fh->warn("ignoring %result before first action");
	return;
    }

    while (defined (my $line = $fh->getline(1))) {
	if ($line =~ /^\s*%/) {
	    $fh->ungetline($line);
	    last;
	}
	$action->add_result($line);
    }
}


# PRIVATE to the parse() method
sub _parse_comment {
    my $this = shift();
    my($fh, $comment) = @_;

    my $action = $this->{_action};
    if (!defined $action) {
	$fh->warn("ignoring %comment before first %action");
	return;
    }

    my $old = $action->comment();
    if (defined $old) {
	$fh->warn("discarding old comment '$old'");
    }

    $action->comment($comment);
}


# PRIVATE to the parse() method
sub _parse_alias {
    my $this = shift();
    my $fh = shift();
    my $type = lc(shift());
    my @words = map { uc() } @_;

    my $href = $this->{$type . 'vocab'};
    for my $word (@words) {
	$fh->fatal("empty word") if !$word;
	push @{ $href->{$word} }, grep { $_ ne $word } @words;
    }
}


# PRIVATE to the parse() method
#
# At this point, we have a bunch of equivalence classes for each
# vocabulary set, we we need to coalesce them.  For example, if we
# have the following in the source --
#	%valias enter go
#	%valias run go
#	%valias walk go
# We'll have a {vvocab} hash that looks like this --
#	RUN -> GO
#	GO -> ENTER,RUN,WALK
#	ENTER -> GO
#	WALK -> GO
# But we want a single list of all four words

sub _coalesce_aliases {
    my $this = shift();

    $this->{vvocab} = _extend_lists($this->{vvocab});
    $this->{nvocab} = _extend_lists($this->{nvocab});
}

# PRIVATE to the _coalesce_aliases() method
sub _extend_lists {
    my($vocab) = @_;

    my @keys = keys %$vocab;	# we're going to change this
    foreach my $key (@keys) {
	#warn "considering aliases for '$key'";
	next if !exists $vocab->{$key};
	#warn "thinking about '$key'";
	my @list = _equivalents($vocab, $key, { $key => 1});
	#warn "\t" . join(' ', @list);
	foreach my $used (@list) {
	    delete $vocab->{$used};
	    #warn "deleted '$used'";
	}
	$vocab->{$key} = [ @list ];
    }

    return $vocab;
}

# PRIVATE to the _extend_lists method()
sub _equivalents {
    my($vocab, $word, $seen) = @_;

    my @equivalents;
    my $cref = $vocab->{$word};
    foreach my $candidate (@$cref) {
	next if exists $seen->{$candidate};
	$seen->{$candidate} = 1;
	my @sub = _equivalents($vocab, $candidate, $seen);
	push @equivalents, $candidate, @sub;
    }

    return @equivalents;
}


1;
